### PR TITLE
Issue#96 configure liquibase migration

### PR DESCRIPTION
**Pull request by Abhai2016**
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

**Review by OlTym — COMMENTED** [review]
In accordance with logical database shema the entity "User" has not field "Contact"

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml [hunk 1]
```diff
@@ -114,10 +85,10 @@
                              primaryKeyName="account_pkey"/>
             </column>
             <column name="CONTACTS" type="VARCHAR">
-                <constraints nullable="false"/>
+                <constraints nullable="true"/>
```
Comment (OlTym) [inline home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml:88]: The  field "CONTACTS"is not in the  entity "User"  in accordance with logical sheme database

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml [hunk 2]
```diff
@@ -114,10 +85,10 @@
...
             </column>
             <column name="CREATE_DATE" type="TIMESTAMP">
-                <constraints nullable="false"/>
+                <constraints nullable="true"/>
```
Comment (OlTym) [inline home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml]: This constraint  must not be  nullable

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml [hunk 3]
```diff
@@ -144,7 +115,7 @@
                 <constraints nullable="false"/>
             </column>
             <column name="UPDATE_DATE" type="TIMESTAMP">
-                <constraints nullable="false"/>
+                <constraints nullable="true"/>
```
Comment (OlTym) [inline home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml]: This constraint must not be nullable

File: README.md [hunk 4]
```diff
@@ -107,22 +107,11 @@ api.secret=${API_SECRET}
 ### How to run tests
 ### How to Checkstyle
 ### How to run data migration
-You can update DB with this CLI application. For that purpose you should run it from home-data-migration directory in two ways.
-1. Use full URL(--url) for connection to your DB
-   
-   `java -jar ./target/home-data-migration-0.0.1-SNAPSHOT.jar --url=jdbc:postgresql://localhost:5432/?user=user&password=password`
-   
-2. Use URL(--url), username(-u or --user) and password(-p or --password) for connection to your DB
```
Comment (PoLytvy) [inline README.md]: According to previous task(33) we had option to run migration with CLI. It is the way how it connects to heroku. Why did you delete it?

File: home-data-migration/pom.xml [hunk 5]
```diff
@@ -55,11 +49,6 @@
     </profiles>
 
     <dependencies>
-        <dependency>
-            <groupId>info.picocli</groupId>
-            <artifactId>picocli</artifactId>
-            <version>${picocli.version}</version>
-        </dependency>
```
Comment (PoLytvy) [inline home-data-migration/pom.xml]: Why did you delete picocli?

File: home-data-migration/src/main/java/com/softserveinc/ita/homeproject/homedatamigration/migrations/Liquibase.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    @Value("${password}")
+    private String password;
+
+    @Value("${changeLogFile}")
+    private String changeLogFile;
+
+
+    @Override
+    public void run(ApplicationArguments args) throws Exception {
+        System.out.println("Liquibase-update \n version = 0.0.1 \n " +
```
Comment (PoLytvy) [inline home-data-migration/src/main/java/com/softserveinc/ita/homeproject/homedatamigration/migrations/Liquibase.java]: Why do you use "sout" here? Not logs

File: home-data-migration/src/main/java/com/softserveinc/ita/homeproject/homedatamigration/migrations/Liquibase.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+                "This application updates DB by using liquibase scripts. \" +\n" +
+                "                \"Use full URL or combination of URL, user and password.");
+
+        SpringLiquibase springLiquibase = new SpringLiquibase();
```
Comment (PoLytvy) [inline home-data-migration/src/main/java/com/softserveinc/ita/homeproject/homedatamigration/migrations/Liquibase.java]: Do we need to use Spring in this module?

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml [hunk 8]
```diff
@@ -0,0 +1,111 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+
+    <changeSet id="user" author="Abhai2016">
```
Comment (PoLytvy) [inline home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: Why id is "user"? It should be name of your task

Comment (PoLytvy) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: And the same for all ids

File: home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/repository/UserRepository.java [hunk 9]
```diff
@@ -20,8 +20,8 @@
     /**
      * Method for finding users in database by email
      *
-     * @param email email of the user that is being searched
-     * @return an instance of Optional(User) class
+     * @param email is the email of the user that is being searched
+     * @return returns an instance of Optional(User) class
```
Comment (PoLytvy) [inline home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/repository/UserRepository.java]: I think, "@return" already means return. And we don't need to emphasize it again

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/DataLoader.java [hunk 10]
```diff
@@ -29,6 +30,12 @@
 @Profile("dev")
 public class DataLoader implements ApplicationRunner {
```
Comment (PoLytvy) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/DataLoader.java]: So do we need to use DataLoader?

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml (hunk 1 above)
Comment (Abhai2016) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml:88]: Ok, I'll fix in my task Issue#98-Configure entities.

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml (hunk 2 above)
Comment (Abhai2016) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml]: Done.

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml (hunk 3 above)
Comment (Abhai2016) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml]: Done.

File: README.md (hunk 4 above)
Comment (Abhai2016) [reply README.md]: Sorry, missed that. Want to discuss it and will return it in needed.

File: home-data-migration/pom.xml (hunk 5 above)
Comment (Abhai2016) [reply home-data-migration/pom.xml]: Sorry, missed that. Want to discuss it and will return it in needed.

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml (hunk 8 above)
Comment (Abhai2016) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: Sorry, didn't know how to name them right. Done.

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/DataLoader.java (hunk 10 above)
Comment (Abhai2016) [reply home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/DataLoader.java]: No, I already fixed that and got rid of DataLoader.

File: home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/repository/UserRepository.java (hunk 9 above)
Comment (Abhai2016) [reply home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/repository/UserRepository.java]: Ok, done.

File: README.md [hunk 11]
```diff
@@ -107,22 +107,11 @@ api.secret=${API_SECRET}
...
+You can update DB with maven plugin. For that purpose you should run it in home-data-migration module by using commands:
```
Comment (kh0ma) [inline README.md]: We have two options:

1. using maven build and enable profile, suitable for local needs
2. using app, suitable for real envs

File: home-application/pom.xml [hunk 12]
```diff
@@ -49,23 +49,6 @@
             <artifactId>home-generated-server</artifactId>
             <version>${revision}</version>
         </dependency>
-        <dependency>
-            <groupId>org.springframework.boot</groupId>
-            <artifactId>spring-boot-devtools</artifactId>
```
Comment (kh0ma) [inline home-application/pom.xml:54]: Do we have it somewhere in another place? Where? Seems to me it is a much preferable place.

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/NewsApiServiceImpl.java [hunk 13]
```diff
@@ -118,4 +118,4 @@ public Response updateNews(Long id, UpdateNews updateNews) {
         return Response.ok().entity(response).build();
     }
 
-}
+}
```
Comment (kh0ma) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/NewsApiServiceImpl.java]: Please do not forget to add a newline at the end of the file.

File: home-data-migration/pom.xml [hunk 14]
```diff
@@ -32,12 +31,7 @@
                         <artifactId>liquibase-maven-plugin</artifactId>
                         <version>${liquibase.version}</version>
                         <configuration>
-                            <promptOnNonLocalDatabase>false</promptOnNonLocalDatabase>
-                            <changeLogFile>db/changelog/master.xml</changeLogFile>
-                            <driver>org.postgresql.Driver</driver>
-                            <url>jdbc:postgresql://localhost:5433/postgres</url>
-                            <username>postgres</username>
-                            <password>password</password>
+                            <propertyFile>config/liquibase.properties</propertyFile>
```
Comment (kh0ma) [inline home-data-migration/pom.xml]: I don't have a strong opinion on it but I think keeping these properties in maven config much preferable. Just move it to the properties section so anybody can override it: `mvn clean install -Dproperty.to.override`

File: home-data-migration/pom.xml [hunk 15]
```diff
@@ -90,7 +75,7 @@
                     <archive>
                         <manifest>
                             <addClasspath>true</addClasspath>
-                            <mainClass>com.softserveinc.ita.homeproject.homedatamigration.migrations.LiquibaseRunner</mainClass>
+                            <mainClass>com.softserveinc.ita.homeproject.homedatamigration.migrations.Liquibase</mainClass>
```
Comment (kh0ma) [inline home-data-migration/pom.xml]: Why did you rename it? What is the purpose?

File: home-data-migration/src/main/java/com/softserveinc/ita/homeproject/homedatamigration/migrations/Liquibase.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.boot.ApplicationRunner;
+import org.springframework.boot.jdbc.DataSourceBuilder;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.stereotype.Component;
+
+import javax.sql.DataSource;
+
+@Component
+@PropertySource("classpath:config/liquibase.properties")
+public class Liquibase implements ApplicationRunner {
```
Comment (kh0ma) [inline home-data-migration/src/main/java/com/softserveinc/ita/homeproject/homedatamigration/migrations/Liquibase.java]: Picocli is much preferable.

File: home-data-migration/src/main/resources/config/liquibase.properties [hunk 17]
```diff
@@ -0,0 +1,8 @@
+changeLogFile=db/changelog/master.xml
+driver=org.postgresql.Driver
+url=jdbc:postgresql://localhost:5433/postgres
+username=postgres
+password=password
+promptOnNonLocalDatabase=false
+parameter.admin.email=admin@example.com
```
Comment (kh0ma) [inline home-data-migration/src/main/resources/config/liquibase.properties]: Let's simplify the task for now and get rid of the idea to provide admin creds. Use some defaults.

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: README.md (hunk 11 above)
Comment (Abhai2016) [reply README.md]: Got back old realization.

File: home-application/pom.xml (hunk 12 above)
Comment (Abhai2016) [reply home-application/pom.xml:54]: It was mentioned twice so I left only one.

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/NewsApiServiceImpl.java (hunk 13 above)
Comment (Abhai2016) [reply home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/NewsApiServiceImpl.java]: Ok, will try to remember that.

File: home-data-migration/src/main/resources/config/liquibase.properties (hunk 17 above)
Comment (Abhai2016) [reply home-data-migration/src/main/resources/config/liquibase.properties]: Done.

File: home-data-migration/src/main/java/com/softserveinc/ita/homeproject/homedatamigration/migrations/Liquibase.java (hunk 16 above)
Comment (Abhai2016) [reply home-data-migration/src/main/java/com/softserveinc/ita/homeproject/homedatamigration/migrations/Liquibase.java]: Yeah, I returned old realization.

File: home-data-migration/pom.xml (hunk 14 above)
Comment (Abhai2016) [reply home-data-migration/pom.xml]: Done.

File: home-data-migration/src/main/java/com/softserveinc/ita/homeproject/homedatamigration/migrations/Liquibase.java (hunk 6 above)
Comment (Abhai2016) [reply home-data-migration/src/main/java/com/softserveinc/ita/homeproject/homedatamigration/migrations/Liquibase.java]: Never mind, I already got back old realization.

File: home-application/pom.xml (hunk 12 above)
Comment (kh0ma) [reply home-application/pom.xml:54]: Since home-application is the module that produces an executable jar the dependency `spring-boot-devtools` should be here. Please remove it from the duplicated place and leave it here.

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/NewsApiServiceImpl.java (hunk 13 above)
Comment (kh0ma) [reply home-application/src/main/java/com/softserveinc/ita/homeproject/application/apiservice/NewsApiServiceImpl.java]: Please add newline at the end of the file.

File: home-data-migration/pom.xml [hunk 18]
```diff
@@ -32,12 +32,11 @@
                         <artifactId>liquibase-maven-plugin</artifactId>
                         <version>${liquibase.version}</version>
                         <configuration>
-                            <promptOnNonLocalDatabase>false</promptOnNonLocalDatabase>
                             <changeLogFile>db/changelog/master.xml</changeLogFile>
-                            <driver>org.postgresql.Driver</driver>
```
Comment (kh0ma) [inline home-data-migration/pom.xml:37]: Is it correct that we removed `<driver>org.postgresql.Driver</driver>`?

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+            <column name="enabled" value="true"/>
+            <column name="update_date" value="2021-01-13 11:30:00"/>
+        </insert>
+    </changeSet>
+
+    <changeSet id="Issue96.002" author="Abhai2016">
+        <insert  dbms="postgresql"
+                 schemaName="public"
+                 tableName="account">
+            <column name="id" value="2"/>
```
Comment (kh0ma) [inline home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: We do not need a user added as liquibase migration. Admin will be enough. Please remove a user.

To be clear we can create any user if we have a user in admin role.

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: home-application/pom.xml (hunk 12 above)
Comment (Abhai2016) [reply home-application/pom.xml:54]: In my branch here there's still spring-boot-devtools.

File: home-data-migration/pom.xml (hunk 18 above)
Comment (Abhai2016) [reply home-data-migration/pom.xml:37]: I guess because it needed for Springliquibase which already I changed on old realization because now it works fine without it.

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml (hunk 19 above)
Comment (Abhai2016) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: Done.

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+        </insert>
+        <insert dbms="postgresql"
+                schemaName="public"
+                tableName="permission">
+            <column name="id" value="9"/>
+            <column name="name" value="DELETE_NEWS_PERMISSION"/>
+        </insert>
+    </changeSet>
+
+    <changeSet id="Issue96.005" author="Abhao2016">
```
Comment (kh0ma) [inline home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: Typo here
author="Abhai2016"

File: home-application/src/main/resources/config/application.properties [hunk 21]
```diff
@@ -1,4 +1,3 @@
-#spring.profiles.active=dev
 management.server.port=${MANAGEMENT_SERVER_PORT:8090}
 management.endpoints.web.exposure.include=health
 management.endpoint.health.show-details=always
```
Comment (kh0ma) [inline home-application/src/main/resources/config/application.properties:3]: Please add this property:
`spring.jpa.hibernate.ddl-auto=validate`

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml (hunk 20 above)
Comment (Abhai2016) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: Done.

File: home-application/src/main/resources/config/application.properties (hunk 21 above)
Comment (Abhai2016) [reply home-application/src/main/resources/config/application.properties:3]: Done.

**Review by kh0ma — APPROVED** [review]

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+            <column name="last_name" value="User"/>
+            <column name="expired" value="false"/>
+            <column name="locked" value="false"/>
+            <column name="credentials_expired" value="false"/>
+            <column name="enabled" value="true"/>
+            <column name="update_date" value="2021-01-13 11:30:00"/>
+        </insert>
+    </changeSet>
+
+    <changeSet id="Issue96.003" author="Abhai2016">
```
Comment (PoLytvy) [inline home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: You have missed "Issue96.002"

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+    <changeSet id="Issue96.005" author="Abhai2016">
+        <insert  dbms="postgresql"
+                 schemaName="public"
+                 tableName="account_role">
+            <column name="role_id" value="1"/>
+            <column name="account_id" value="1"/>
+        </insert>
+    </changeSet>
+
+    <changeSet id="role_permission" author="Abhai2016">
```
Comment (PoLytvy) [inline home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: Please, use the same naming ways for all changesets ids

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml [hunk 24]
```diff
@@ -5,35 +5,6 @@
...
-            </column>
-            <column name="flat" type="VARCHAR(255)">
-                <constraints nullable="false"/>
-            </column>
-            <column name="active" type="boolean">
-                <constraints nullable="false"/>
-            </column>
-        </createTable>
-    </changeSet>
-
```
Comment (PoLytvy) [inline home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml:36]: Why do you delete this table?

File: home-data-migration/pom.xml (hunk 18 above)
Comment (PoLytvy) [reply home-data-migration/pom.xml:37]: I think, it should be added back. Because you may have this driver installed in your computer, but probably it can provide problems on others

**Review by OlTym — APPROVED** [review]

**Review by PoLytvy — CHANGES_REQUESTED** [review]

**Review by Romanov-Niko — APPROVED** [review]

File: home-data-migration/pom.xml (hunk 18 above)
Comment (Abhai2016) [reply home-data-migration/pom.xml:37]: Done.

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml (hunk 22 above)
Comment (Abhai2016) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: Done.

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml (hunk 23 above)
Comment (Abhai2016) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue96.xml]: Done.

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml (hunk 24 above)
Comment (Abhai2016) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue33.xml:36]: Because we are not gonna use it so it's dead code as Aleksandr called it.

**Review by PoLytvy — APPROVED** [review]
